Sources/Testing/Polling/Polling.swift

```diff
@@ -0,0 +1,293 @@
+  ///
+  /// - Parameters:
+  ///   - behavior: The PollingBehavior to use.
+  ///   - timeout: How long to poll for until we time out.
+  ///   - closure: The closure to continuously evaluate.
+  ///   - expression: The expression, corresponding to `condition`, that is being
+  ///     evaluated (if available at compile time.)
+  ///   - comments: An array of comments describing the expectation. This array
+  ///     may be empty.
+  ///   - isRequired: Whether or not the expectation is required. The value of
+  ///     this argument does not affect whether or not an error is thrown on
+  ///     failure.
+  ///   - sourceLocation: The source location of the expectation.
+  static func run<R>(
+    behavior: PollingBehavior,
+    timeout: Duration,
+    closure: @escaping @Sendable () async -> R?,
+    expression: __Expression,
+    comments: [Comment],
+    isRequired: Bool,
+    sourceLocation: SourceLocation
+  ) async -> Result<R, any Error> where R: Sendable {
+    var expectation = Expectation(
+      evaluatedExpression: expression,
+      isPassing: true,
+      isRequired: isRequired,
+      sourceLocation: sourceLocation
+    )
+
+    let recorder = Recorder<R>()
+
+    let result = await poll(expression: {
+      if let value = await closure() {
+        await recorder.record(value: value)
+        return true
+      }
+      return false
+    }, behavior: behavior, timeout: timeout)
+
+    let sourceContext = SourceContext(backtrace: nil, sourceLocation: sourceLocation)
+
+    switch result {
+    case .timedOut:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .timedOutWithoutRunning:
+      expectation.isPassing = false
+      Issue(
+        kind: .expectationFailed(expectation),
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    case .finished:
+      return __checkValue(
+        await recorder.lastValue,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .failed:
+      return __checkValue(
+        nil,
+        expression: expression,
+        comments: comments,
+        isRequired: isRequired,
+        sourceLocation: sourceLocation
+      )
+    case .cancelled:
+      Issue(
+        kind: .system,
+        comments: comments,
+        sourceContext: sourceContext
+      ).record()
+    }
+
+    return .failure(ExpectationFailedError(expectation: expectation))
+  }
+
+  /// A type to record the last value returned by a closure returning an optional
+  /// This is only used in the `#require(until:)` macro returning an optional.
+  private actor Recorder<R: Sendable> {
+    var lastValue: R?
+
+    /// Record a new value to be returned
+    func record(value: R) {
+      self.lastValue = value
+    }
+  }
```

Comment:
Ugh. I really dislike this approach, but I wanted to get this PR up so that others could actually test this proposal out & I couldn't think of a better way to handle this. Hopefully one of the benefits of a non-timed-based approach to polling will be that returning values will be much simpler/not a tacked-on hack.